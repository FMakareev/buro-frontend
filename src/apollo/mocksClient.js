import faker from 'faker';
import setupClient from './helpers/apolloClientMock';
import schema from './schema.graphqls';
import { userDocumentList } from './graphql/query/userDocumentList';
import { userList } from './graphql/query/userList';
import { userItem } from './graphql/query/userItem';
import { notificationList } from './graphql/query/notificationList';
import { ROLE_BANK, ROLE_BUREAU, ROLE_CLIENT } from '../shared/roles';

const defaultMocks = {
  Query: () => ({
    userList,
    userItem,
    userEmailItem: (query, { email }) => {
      console.log(query, email);
      switch (email) {
        case 'client@test.com': {
          return {
            ...userItem(),
            email: 'client@test.com',
            role: ROLE_CLIENT,
          };
        }
        case 'bank@test.com': {
          return {
            ...userItem(),
            email: 'bank@test.com',
            role: ROLE_BANK,
          };
        }
        case 'bureau@test.com': {
          return {
            ...userItem(),
            email: 'bureau@test.com',
            role: ROLE_BUREAU,
          };
        }
        default: {
          throw Error(
            JSON.stringify({
              userEmailItem: null,
              errors: [
                {
                  message: 'GraphQL error: user not found',
                },
              ],
            }),
          );
        }
      }
    },
    userDocumentList,
    notificationList,
  }),
  Mutation: () => ({
    /**
     * @params {func} mutation - этот же запрос
     * @params {object} props - аргументы которые были переданы
     * */
    createUser: (mutation, props) => props,
    updateUser: (mutation, props) => props,
    userResetPassword: (mutation, props) => props,
    createNotification: (mutation, props) =>
      // для имитации запроса к серверу с рандомной задержкой и результатом.
      new Promise((resolve, reject) => {
        setTimeout(() => {
          faker.random.number(1)
            ? resolve(JSON.stringify({ data: { createNotification: props } }))
            : reject(
                JSON.stringify({
                  errors: [
                    {
                      message: 'error!',
                    },
                  ],
                }),
              );
        }, faker.random.number(2000));
      }),
    updateNotification: (mutation, props) =>
      // для имитации запроса к серверу с рандомной задержкой и результатом.
      new Promise((resolve, reject) => {
        setTimeout(() => {
          faker.random.number(1)
            ? resolve(JSON.stringify({ data: { updateNotification: props } }))
            : reject(
                JSON.stringify({
                  errors: [
                    {
                      message: 'error!',
                    },
                  ],
                }),
              );
        }, faker.random.number(2000));
      }),
    userPasswordRecovery: (mutation, props) =>
      // throw Error(JSON.stringify({
      //   userPasswordRecovery: null,
      //   errors: [
      //     {
      //       message:'Error!'
      //     }
      //   ]
      // }))
      props,
  }),
};

export const mocksClient = setupClient(defaultMocks, schema)();

export default mocksClient;
