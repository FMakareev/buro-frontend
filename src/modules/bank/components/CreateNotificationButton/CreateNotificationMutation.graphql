mutation ($clientid: String, $bankid: String, $status: String){
  createnotification(client: $clientid, bank:$bankid, status:$status) {
    notification {
      status
    }
  }
}