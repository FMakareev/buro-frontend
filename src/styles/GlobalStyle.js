import { createGlobalStyle } from 'styled-components';
import BG from '../assets/image/bg.jpg';

export const GlobalStyle = createGlobalStyle`
  *{
    box-sizing: border-box;
  }
  body, html, #app {
    padding: 0;
    margin: 0;
    width: 100%;
    min-height: 100vh;
    background-color: #fff;
    font-family: 'Montserrat Regular';
    font-size: 16px;
  }
  #app{
    background-color: #fff;
    background-image: url(${BG});
    background-repeat: no-repeat;
    background-position: center;
    background-size: cover;
  }
  a {
    text-decoration: none;
    color: inherit;
  }
  
  button:focus, input:focus, select:focus {
    outline: none !important;
    box-shadow: 0 0 0 5px #E7EDF8;
  }
  

  @font-face {
    font-family: 'Montserrat Regular';
    src: url('../assets/fonts/Montserrat-Regular.ttf') format('truetype');
    font-weight: normal;
    font-style: normal;
  }

  @font-face {
  font-family: 'Montserrat Medium';
  src: url('../assets/fonts/Montserrat-Medium.ttf') format('truetype');
  font-weight: normal;
  font-style: normal;
  }

  @font-face {
  font-family: 'Montserrat Bold';
  src: url('../assets/fonts/Montserrat-Bold.ttf') format('truetype');
  font-weight: bold;
  font-style: normal;
}

`;

export default GlobalStyle;
