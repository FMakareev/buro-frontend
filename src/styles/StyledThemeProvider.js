import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'rebass';
import { ButtonVariant } from './variants/buttonVariant';
import { ButtonSize } from './variants/buttonSize';
import { TextVariant } from './variants/textVariant';

export const ColorPallet = {
  color0: '#FFFFFF',
  color1: '#093971',
  color2: '#EFF9FC',
  color3: '#E7EDF8',
  color4: '#61C3E2',
  color5: '#A2A2A2',
  color6: '#0F60BD',
  color7: '#072C57',
  color8: '#61C3E2',
  color9: '#EB5757',
  color10: '#E7EDF8',
  color11: '',
};

const Space = [
  0,
  2,
  4,
  8,
  12,
  14,
  16,
  18,
  20,
  24,
  28,
  32,
  36,
  40,
  44,
  48,
  52,
  56,
  60,
  64,
  68,
  72,
  76,
  80,
];

const Weight = [300, 500, 700];

const Breakpoints = ['576px', '768px', '992px', '1200px'];

const boxShadow = ['4px 8px 16px rgba(28, 65, 84, 0.08)'];

export const StyledThemeProvider = ({ children }) => (
  <Provider
    theme={{
      space: Space,
      fontSizes: Space,
      fontWeight: Weight,
      lineHeight: Space,
      boxShadow,
      borderRadius: Space,
      borderColor: ColorPallet,
      breakpoints: Breakpoints,
      colors: ColorPallet,
      variant: {
        buttons: ButtonVariant,
        buttonSize: ButtonSize,
        text: TextVariant,
      },
      fontFamily: {
        regular: 'Montserrat Regular',
        medium: 'Montserrat Medium',
        bold: 'Montserrat Bold',
      },
    }}>
    {children}
  </Provider>
);

StyledThemeProvider.propTypes = {
  children: PropTypes.element.isRequired,
};

StyledThemeProvider.defaultProps = {
  children: 'children is empty',
};

export default StyledThemeProvider;
