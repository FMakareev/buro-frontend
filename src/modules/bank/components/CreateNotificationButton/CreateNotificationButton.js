import React from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { ButtonWithImage } from '@lib/ui/ButtonWithImage/ButtonWithImage';
import { Text } from '@lib/ui/Text/Text';
import { SpeedingWheel } from '@lib/ui/SmallPreloader/SmallPreloader';
import CreateNotificationMutation from './CreateNotificationMutation.graphql';


/**
 * @desc Кнопка для вызова мутации создания уведомления к пользователю
 * */
export const CreateNotificationButton = ({ children, client, bank }) => (
  <Mutation mutation={CreateNotificationMutation} onError={()=>{}}>
    {(mutate, { called, data, error, loading }) => {
      /** появляется если called - запрос был вызван, !loading - загрузка не идет, !error - нет ошибок */
      if (called && !loading && !error) {
        return (
          <Text fontSize={6} color="color1">
            Pending approval
          </Text>
        );
      }
      return (
        <ButtonWithImage
          disabled={loading}
          onClick={() => mutate({ variables: { client, bank } })}
          display="inline-block"
          iconRight={
            loading ? (
              <Text fontSize={5} lineHeight={0} fill="inherit">
                <SpeedingWheel />
              </Text>
            ) : null
          }
          size="xsmall"
          testID={'CreateNotificationButton'}
          variant={error ? 'error' : 'transparent'}
          pl="3px"
          pr="5px">
          {children}
        </ButtonWithImage>
      );
    }}
  </Mutation>
);
CreateNotificationButton.propTypes = {
  /** id банка */
  bank: PropTypes.string.isRequired,
  /** id клиента */
  client: PropTypes.string.isRequired,
};
export default CreateNotificationButton;
