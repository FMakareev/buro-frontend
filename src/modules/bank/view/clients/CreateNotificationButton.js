import React from 'react';
import PropTypes from 'prop-types';
import CreateNotificationMutation from './CreateNotificationMutation.graphql';
import {Mutation} from "react-apollo";
import {ButtonWithImage} from "../../../../components/ButtonWithImage/ButtonWithImage";
import {Text} from "../../../../components/Text/Text";
import {SpeedingWheel} from "../../../../components/SmallPreloader/SmallPreloader";

/**
 * @desc Кнопка для вызова мутации создания уведомления к пользователю
 * */
export const CreateNotificationButton = ({children, id}) => {
  return (<Mutation mutation={CreateNotificationMutation}>
      {(createNotification, {called, date, error, loading}) => {
        console.log('Mutation: called', called);
        console.log('Mutation: date', date);
        console.log('Mutation: loading', loading);
        console.log('Mutation: error', error);
        /** появляется если called - запрос был вызван, !loading - загрузка не идет, !error - нет ошибок*/
        if (called && !loading && !error) {
          return (<Text fontSize={6} color={'color1'}>
            Pending approval
          </Text>)
        }
        return (
          <ButtonWithImage
            disabled={loading}
            onClick={() => {
              createNotification({variables: {id}})
            }}
            display={"inline-block"}
            iconRight={
              loading ? <Text fontSize={5} lineHeight={0} fill="inherit">
                <SpeedingWheel/>
              </Text> : null
            }
            size={"xsmall"}
            variant={error ? "error" : "transparent"}
            pl={"3px"}
            pr={"5px"}>
            {children}
          </ButtonWithImage>
        )

      }}
    </Mutation>
  )
};
CreateNotificationButton.propTypes = {
  /** id банка */
  id: PropTypes.string.isRequired,
};
export default CreateNotificationButton;
