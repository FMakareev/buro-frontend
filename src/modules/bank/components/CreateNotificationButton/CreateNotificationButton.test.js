import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import { ApolloProvider } from 'react-apollo';
import { MockedProvider } from 'react-apollo/test-utils';

import wait from 'waait';
import faker from 'faker';
import { MemoryRouter, Redirect } from 'react-router-dom';
import { Provider as ProviderRedux } from 'react-redux';
import configureStore from 'redux-mock-store';
import mocksClient from '../../../../apollo/mocksClient';

import { CreateNotificationButton } from './CreateNotificationButton';
import { StyledThemeProvider } from '../../../../styles/StyledThemeProvider';

import CreateNotificationMutation from './CreateNotificationMutation.graphql';

test('CreateNotificationButton: обычное состояние', () => {
  // const store = mockStore({});

  const output = renderer.create(
    <StyledThemeProvider>
      <MemoryRouter>
        <ApolloProvider client={mocksClient}>
          <CreateNotificationButton>Request</CreateNotificationButton>
        </ApolloProvider>
      </MemoryRouter>
    </StyledThemeProvider>,
  );

  expect(output).toMatchSnapshot();
});

test('CreateNotificationButton: вызов запроса', () => {
  const output = mount(
    <StyledThemeProvider>
      <MemoryRouter>
        <ApolloProvider client={mocksClient}>
          <CreateNotificationButton>Request</CreateNotificationButton>
        </ApolloProvider>
      </MemoryRouter>
    </StyledThemeProvider>,
  );

  output.find('button').simulate('click');

  wait(0);
  expect(output.find('button').props().disabled).toBe(true);
});

test('CreateNotificationButton: запрос завершен', async () => {
  const mocks = [
    {
      request: {
        CreateNotificationMutation,
        variables: {
          bankid: faker.random.uuid(),
          // clientid: faker.random.uuid(),
        },
      },
      result: {
        data: [{ status: 'requested' }],
      },
    },
  ];

  const props = { id: faker.random.uuid() };

  const output = renderer.create(
    <StyledThemeProvider>
      <MemoryRouter>
        <MockedProvider mocks={mocks}>
          <ApolloProvider client={mocksClient}>
            <CreateNotificationButton {...props}>Request</CreateNotificationButton>
          </ApolloProvider>
        </MockedProvider>
      </MemoryRouter>
    </StyledThemeProvider>,
  );

  const button = output.root.findByType('button');
  button.props.onClick();

  await wait(1);

  const tree = output.toJSON();
  expect(tree).toMatchSnapshot();
});

test('CreateNotificationButton: запрос завершен ошибкой', async () => {
  const mocks = [
    {
      request: {
        CreateNotificationMutation,
        variables: {
          bankid: faker.random.uuid(),
          // clientid: faker.random.uuid(),
        },
      },
      error: new Error('Network Error!'),
      result: {
        errors: [{ message: 'GraphQLError!' }],
      },
    },
  ];

  const props = { id: faker.random.uuid() };

  const output = renderer.create(
    <StyledThemeProvider>
      <MemoryRouter>
        <MockedProvider mocks={mocks} addTypename>
          <ApolloProvider client={mocksClient}>
            <CreateNotificationButton {...props}>Request</CreateNotificationButton>
          </ApolloProvider>
        </MockedProvider>
      </MemoryRouter>
    </StyledThemeProvider>,
  );

  const button = output.root.findByType('button');
  button.props.onClick();

  await wait(1);

  const tree = output.toJSON();
  expect(tree).toMatchSnapshot();
});
