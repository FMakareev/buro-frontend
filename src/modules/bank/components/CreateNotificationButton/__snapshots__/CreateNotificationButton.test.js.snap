// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNotificationButton: запрос завершен 1`] = `"<div class=\\"sc-bxivhb dUiVDW\\"><div font-size=\\"6\\" color=\\"color1\\" class=\\"sc-1jxwge6-0 sc-18ch56c-0 hxjSXI\\">Pending approval</div></div>"`;

exports[`CreateNotificationButton: запрос завершен ошибкой 1`] = `"<div class=\\"sc-bxivhb dUiVDW\\"><div font-size=\\"6\\" color=\\"color1\\" class=\\"sc-1jxwge6-0 sc-18ch56c-0 hxjSXI\\">Pending approval</div></div>"`;

exports[`CreateNotificationButton: запрос завершен ошибкой 2`] = `"<div class=\\"sc-bxivhb dUiVDW\\"><div font-size=\\"6\\" color=\\"color1\\" class=\\"sc-1jxwge6-0 sc-18ch56c-0 hxjSXI\\">Pending approval</div></div>"`;

exports[`CreateNotificationButton: обычное состояние 1`] = `
<div
  className="sc-bxivhb dUiVDW"
>
  <button
    className="mxzgdr-0 bYJcNk"
    disabled={false}
    display="inline-block"
    onClick={[Function]}
    size="xsmall"
    type="button"
  >
    <div
      className="sc-1jxwge6-0 wterhv-0 eZGOHC"
      width="100%"
    >
      <div />
      <div
        className="sc-1jxwge6-0 wterhv-0 irkFkV"
      >
        Request
      </div>
      <div />
    </div>
  </button>
</div>
`;
